govcd/landing_zone_add_on.go

```diff
@@ -0,0 +1,492 @@
+	if err != nil {
+		return nil, err
+	}
+
+	entityCfg := &types.DefinedEntity{
+		EntityType: fmt.Sprintf("urn:vcloud:type:%s:%s:%s", slzAddOnRdeType[0], slzAddOnRdeType[1], slzAddOnRdeType[2]),
+		Name:       rdeName,
+		State:      addrOf("PRE_CREATED"),
+		Entity:     unmarshalledRdeEntityJson,
+	}
+
+	createdRdeEntity, err := rdeType.CreateRde(*entityCfg, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RDE entity: %s", err)
+	}
+
+	err = createdRdeEntity.Resolve()
+	if err != nil {
+		return nil, fmt.Errorf("error resolving Solutions Add-On after creating: %s", err)
+	}
+
+	result, err := convertRdeEntityToAny[types.SolutionAddOn](createdRdeEntity.DefinedEntity.Entity)
+	if err != nil {
+		return nil, err
+	}
+
+	returnType := SolutionAddOn{
+		SolutionAddOnEntity: result,
+		vcdClient:           vcdClient,
+		DefinedEntity:       createdRdeEntity,
+	}
+
+	return &returnType, nil
+}
+
+// GetAllSolutionAddons retrieves all Solution Add-Ons with a given filter
+func (vcdClient *VCDClient) GetAllSolutionAddons(queryParameters url.Values) ([]*SolutionAddOn, error) {
+	allAddons, err := vcdClient.GetAllRdes(slzAddOnRdeType[0], slzAddOnRdeType[1], slzAddOnRdeType[2], queryParameters)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving all Solution Add-ons: %s", err)
+	}
+
+	results := make([]*SolutionAddOn, len(allAddons))
+	for index, rde := range allAddons {
+		addon, err := convertRdeEntityToAny[types.SolutionAddOn](rde.DefinedEntity.Entity)
+		if err != nil {
+			return nil, fmt.Errorf("error converting RDE to Solution Add-on: %s", err)
+		}
+
+		results[index] = &SolutionAddOn{
+			vcdClient:           vcdClient,
+			DefinedEntity:       rde,
+			SolutionAddOnEntity: addon,
+		}
+	}
+
+	return results, nil
+}
+
+// GetSolutionAddonById retrieves Solution Add-On by ID
+func (vcdClient *VCDClient) GetSolutionAddonById(id string) (*SolutionAddOn, error) {
+	if id == "" {
+		return nil, fmt.Errorf("id must be specified")
+	}
+	rde, err := getRdeById(&vcdClient.Client, id)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving Solution Add-On by ID: %s", err)
+	}
+
+	result, err := convertRdeEntityToAny[types.SolutionAddOn](rde.DefinedEntity.Entity)
+	if err != nil {
+		return nil, err
+	}
+
+	packages := &SolutionAddOn{
+		SolutionAddOnEntity: result,
+		vcdClient:           vcdClient,
+		DefinedEntity:       rde,
+	}
+
+	return packages, nil
+}
+
+// GetSolutionAddonByName retrieves Solution Add-Ons by name
+// Example name: "vmware.ds-1.4.0-23376809"
+func (vcdClient *VCDClient) GetSolutionAddonByName(name string) (*SolutionAddOn, error) {
+	if name == "" {
+		return nil, fmt.Errorf("name must be specified")
+	}
+
+	queryParams := url.Values{}
+	queryParams.Add("filter", fmt.Sprintf("name==%s", name))
+	results, err := vcdClient.GetAllSolutionAddons(queryParams)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving Solution Add-Ons: %s", err)
+	}
+
+	return oneOrError("name", name, results)
```

Comment:
Dumb question, can names be repeated? (It happened with K8s clusters)
And in that case, shall we return a slice?

Reply:
Not really. The `name` itself is quite stupid in this case and depending on Add-On version (`vmware.ds-1.4.0-23376809`). One can't upload two Add-Ons of the same version as far as I've experimented